h/traversal/group_membership.py

```diff
@@ -0,0 +1,51 @@
+from dataclasses import dataclass
+
+from pyramid.httpexceptions import HTTPNotFound
+
+from h.exceptions import InvalidUserId
+from h.models import Group, GroupMembership, User
+
+
+@dataclass
+class GroupMembershipContext:
+    group: Group
+    user: User
+    membership: GroupMembership | None
+
+
+def group_membership_api_factory(request) -> GroupMembershipContext:
+    user_service = request.find_service(name="user")
+    group_service = request.find_service(name="group")
+    group_members_service = request.find_service(name="group_members")
+
+    userid = request.matchdict["userid"]
+    pubid = request.matchdict["pubid"]
+
+    def get_user() -> User | None:
+        if userid == "me":
+            if request.authenticated_userid:
+                return user_service.fetch(request.authenticated_userid)
+
+            return None
```

Comment:
The API supports `me` as an alias for your own user ID. On `main` `me` is [handled in the view directly](https://github.com/hypothesis/h/blob/7f53416154863cc080a3fb520bbe8b3324409b53/h/views/api/groups.py#L166-L170) but we're going to want this `me` support in the upcoming `PATCH` API as well (and also in a `GET` API if we ever add one), moving it into the context factory here means that all the views support it for free.

Reply:
Technically this means these four lines of code for supporting the `me` alias are now being executed for the add-group-member API as well, no longer just the remove-group-member API. But this has no effect: the add-group-member API only works with authclient authentication never with user authentication (and it doesn't support `X-Forwarded-User`) so when doing `POST /api/group/{pubid}/members/me` `request.authenticated_userid` here will be `None` and this context factory function will raise `HTTPNotFound` which will give the same response as on `main`: 404.